File: sdks/rust/src/pubsub.rs
//! Pub/Sub operations

use crate::client::SynapClient;
use crate::error::Result;
use serde_json::{Value, json};
use std::collections::HashMap;

/// Pub/Sub Manager interface
///
/// Uses StreamableHTTP protocol for all operations.
/// Pub/Sub is **reactive by default** - use `subscribe()` and `subscribe_topic()`.
#[derive(Clone)]
pub struct PubSubManager {
    pub(crate) client: SynapClient,
}

impl PubSubManager {
    /// Create a new Pub/Sub manager interface
    pub(crate) fn new(client: SynapClient) -> Self {
        Self { client }
    }

    /// Publish a message to a topic
    ///
    /// # Returns
    /// Returns the number of subscribers that received the message
    ///
    /// # Example
    /// ```no_run
    /// # use synap_sdk::{SynapClient, SynapConfig};
    /// # use serde_json::json;
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), Box<dyn std::error::Error>> {
    /// # let client = SynapClient::new(SynapConfig::new("http://localhost:15500"))?;
    /// let count = client.pubsub().publish(
    ///     "user.created",
    ///     json!({"id": 123, "name": "Alice"}),
    ///     Some(5),
    ///     None
    /// ).await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn publish(
        &self,
        topic: &str,
        data: Value,
        priority: Option<u8>,
        headers: Option<HashMap<String, String>>,
    ) -> Result<usize> {
        let mut payload = json!({
            "topic": topic,
            "payload": data,  // ✅ FIX: Use "payload" instead of "data" to match server API
        });

        if let Some(p) = priority {
            payload["priority"] = json!(p);
        }

        if let Some(h) = headers {
            payload["headers"] = json!(h);
        }

        let response = self.client.send_command("pubsub.publish", payload).await?;

        Ok(response["subscribers_matched"].as_u64().unwrap_or(0) as usize)
    }

    /// Subscribe to topics
    ///
    /// Supports wildcard patterns:
    /// - `user.*` - single-level wildcard
    /// - `user.#` - multi-level wildcard
    ///
    /// # Returns
    /// Returns a subscription ID
    pub async fn subscribe_topics(
        &self,
        subscriber_id: &str,
        topics: Vec<String>,
    ) -> Result<String> {
        let payload = json!({
            "subscriber_id": subscriber_id,
            "topics": topics,
        });

        let response = self
            .client
            .send_command("pubsub.subscribe", payload)
            .await?;

        Ok(response["subscription_id"]
            .as_str()
            .unwrap_or_default()
            .to_string())
    }

    /// Unsubscribe from topics
    pub async fn unsubscribe(&self, subscriber_id: &str, topics: Vec<String>) -> Result<()> {
        let payload = json!({
            "subscriber_id": subscriber_id,
            "topics": topics,
        });

        self.client
            .send_command("pubsub.unsubscribe", payload)
            .await?;
        Ok(())
    }

    /// List all active topics
    pub async fn list_topics(&self) -> Result<Vec<String>> {
        let response = self.client.send_command("pubsub.topics", json!({})).await?;
        Ok(serde_json::from_value(response["topics"].clone())?)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::SynapConfig;

    #[test]
    fn test_pubsub_manager_creation() {
        let config = SynapConfig::new("http://localhost:15500");
        let client = SynapClient::new(config).unwrap();
        let pubsub = client.pubsub();

        assert!(std::mem::size_of_val(&pubsub) > 0);
    }

    #[test]
    fn test_pubsub_manager_clone() {
        let config = SynapConfig::new("http://localhost:15500");
        let client = SynapClient::new(config).unwrap();
        let pubsub1 = client.pubsub();
        let pubsub2 = pubsub1.clone();

        assert!(std::mem::size_of_val(&pubsub1) > 0);
        assert!(std::mem::size_of_val(&pubsub2) > 0);
    }
}

File: sdks/typescript/src/pubsub.ts
/**
 * Synap TypeScript SDK - Pub/Sub Module
 * 
 * Pub/Sub operations with reactive subscription patterns.
 */

import { Observable, Subject } from 'rxjs';
import {  
  share,
  takeUntil,
} from 'rxjs/operators';
import type { SynapClient } from './client';
import type {
  PubSubPublishOptions,
  PubSubSubscriberOptions,
  ProcessedPubSubMessage,
} from './types';

/**
 * Pub/Sub Manager with reactive support
 */
export class PubSubManager {
  private subscriptions = new Map<string, Subject<void>>();

  constructor(private client: SynapClient) {}

  /**
   * Publish a message to a topic
   */
  async publish(
    topic: string,
    data: any,
    options?: PubSubPublishOptions
  ): Promise<boolean> {
    const cmdPayload: Record<string, any> = {
      topic,
      payload: data,  // ✅ FIX: Use "payload" instead of "data" to match server API
    };

    if (options?.priority !== undefined) {
      cmdPayload.priority = options.priority;
    }

    if (options?.headers) {
      cmdPayload.headers = options.headers;
    }

    const result = await this.client.sendCommand<{ message_id: string; subscribers_matched: number }>(
      'pubsub.publish',
      cmdPayload
    );

    // Return true if publish succeeded (message_id present)
    return !!result.message_id;
  }

  /**
   * Helper: Publish a typed message
   */
  async publishMessage<T>(
    topic: string,
    data: T,
    options?: PubSubPublishOptions
  ): Promise<boolean> {
    return this.publish(topic, data, options);
  }

  // ==================== REACTIVE METHODS ====================

  /**
   * Create a reactive pub/sub subscriber as an Observable
   * 
   * Note: This method creates a simulated reactive subscription using polling.
   * For real-time WebSocket-based subscriptions, consider using the WebSocket API directly.
   * 
   * @example
   * ```typescript
   * synap.pubsub.subscribe({
   *   topics: ['user.created', 'user.updated'],
   *   subscriberId: 'subscriber-1'
   * }).subscribe({
   *   next: (message) => {
   *     console.log('Topic:', message.topic);
   *     console.log('Data:', message.data);
   *   }
   * });
   * ```
   */
  subscribe<T = any>(options: PubSubSubscriberOptions): Observable<ProcessedPubSubMessage<T>> {
    const {
      topics,
      subscriberId = `subscriber-${Date.now()}`,
    } = options;

    const stopKey = `${subscriberId}:${topics.join(',')}`;
    const stopSignal = new Subject<void>();
    this.subscriptions.set(stopKey, stopSignal);

    // Create observable for pub/sub messages
    // This is a simplified implementation using Subject
    // In a real implementation, this would connect to WebSocket
    const messageSubject = new Subject<ProcessedPubSubMessage<T>>();

    // Setup connection (simulated for now)
    // In production, this would open a WebSocket connection
    const setupSubscription = async () => {
      try {
        // Subscribe to topics
        await this.client.sendCommand('pubsub.subscribe', {
          topics,
          subscriber_id: subscriberId,
        });
      } catch (error) {
        console.error(`Error subscribing to topics ${topics.join(', ')}:`, error);
        messageSubject.error(error);
      }
    };

    // Start subscription
    setupSubscription();

    return messageSubject.pipe(
      takeUntil(stopSignal),
      share()
    ) as Observable<ProcessedPubSubMessage<T>>;
  }

  /**
   * Subscribe to a single topic with reactive pattern
   * 
   * @example
   * ```typescript
   * synap.pubsub.subscribeTopic('user.created').subscribe({
   *   next: (message) => console.log('User created:', message.data)
   * });
   * ```
   */
  subscribeTopic<T = any>(topic: string, subscriberId?: string): Observable<ProcessedPubSubMessage<T>> {
    return this.subscribe<T>({
      topics: [topic],
      subscriberId,
    });
  }

  /**
   * Stop a reactive subscription
   * 
   * @param subscriberId - Subscriber ID
   * @param topics - Topics that were subscribed to
   */
  unsubscribe(subscriberId: string, topics: string[]): void {
    const stopKey = `${subscriberId}:${topics.join(',')}`;
    const stopSignal = this.subscriptions.get(stopKey);
    
    if (stopSignal) {
      stopSignal.next();
      stopSignal.complete();
      this.subscriptions.delete(stopKey);
    }

    // Unsubscribe from server
    this.client.sendCommand('pubsub.unsubscribe', {
      subscriber_id: subscriberId,
      topics,
    }).catch((error) => {
      console.error(`Error unsubscribing from topics:`, error);
    });
  }

  /**
   * Stop all reactive subscriptions
   */
  unsubscribeAll(): void {
    this.subscriptions.forEach((signal) => {
      signal.next();
      signal.complete();
    });
    this.subscriptions.clear();
  }

  /**
   * Get statistics for a topic (if supported by server)
   */
  async stats(topic: string): Promise<any> {
    return this.client.sendCommand('pubsub.stats', {
      topic,
    });
  }

  /**
   * List all active topics
   */
  async listTopics(): Promise<string[]> {
    const result = await this.client.sendCommand<{ topics: string[] }>('pubsub.list', {});
    return result.topics;
  }
}

